### ISSUE TITLE
A way to run the "update" stage potentially many times per render?

### ISSUE THREAD

**Issue by rodya-mirov** [bevyengine/bevy#3600]
## What problem does this solve or what need does it fill?

I'm writing a game which is turn-based, with one player and (potentially many) NPC "players." In terms of control flow, think Civilization (but simpler), so a player's turn actions may impact what the next player does. But, in Civ, an NPC civilization will often take a full second to take their turn; while in my game a turn will usually take much less than a millisecond. So I'd like to be able to process as many as possible per tick and give control back to the player as quickly as possible so they experience smooth gameplay in the "normal" case.

## What solution would you like?

The most obvious solution is to somehow run a stage (or system set) "as many times as possible" in the tick, then just render whatever's there. There may be better solutions; I'm open to ideas.

## What alternative(s) have you considered?

Nothing great.

Most obviously I tried a proper turn-based system -- the entity whose turn it is can take an action (or not), then a bunch of followup systems run, then it passes control to the next person at the end. I don't see a way to make that _not_ be the entire update stage, so this is the whole stage, and so if you have (e.g.) 1 player and 59 NPCs then the player can take exactly 1 action per second, which doesn't feel great. At 500 NPCs it would be completely unplayable, so it starts to inform game design in a bad way.

I've also tried writing the code in a way that handles everybody's turn in one tick, but when I have NPCs in a loop, the nice separation of concerns ECS provides go out the window -- I really need [NPC 1]'s actions to be fully resolved before NPC 2 even considers their choices. So I can't launch an event "NPC 1 chooses this" and have a bunch of followup systems run afterward, before "NPC 2 chooses this" occurs.

I've also tried just not caring about where (e.g.) NPC 1's actions may invalidate NPC 2's actions, but this is starting to drive my game design ("oh, this is going to make bugs, let's do something else") in a way that I don't like.

## Additional context

Nothing obvious comes to mind

**Comment by alice-i-cecile** [comment]
So, the best way to do this right now is to make an exclusive system that runs a schedule in a tight loop. Add your game logic systems to this schedule, rather than the schedule owned by the `App`. After each pass of the loop, check if the elapsed time has exceeded your time budget, and break back into the App's schedule if it has.

This will do exactly what you need, and is actually quite ergonomic and powerful!

The fact that you can do this is not at all obvious however, nor is it properly documented in an example. IMO we should have a game example that does this, but I'm struggling to think of a simple game where this pattern would fit well.

**Comment by rodya-mirov** [comment]
@alice-i-cecile Does this still work in bevy 0.6? I don't see any official documentation on "exclusive systems", but according to the bevy cheatbook [link](https://bevy-cheatbook.github.io/cheatsheet.html?highlight=exclusive#bevy-programming-jargon) this is just a system which takes a world as an argument. In fact there's an example in the official repo which does this -- the [ecs-guide](https://github.com/bevyengine/bevy/blob/latest/examples/ecs/ecs_guide.rs) whose code looks like this:

```
// If you really need full, immediate read/write access to the world or resources, you can use a
// "thread local system". These run on the main app thread (hence the name "thread local")
// WARNING: These will block all parallel execution of other systems until they finish, so they
// should generally be avoided if you care about performance
#[allow(dead_code)]
fn thread_local_system(world: &mut World) {
    // this does the same thing as "new_player_system"
    let total_players = world.get_resource_mut::<GameState>().unwrap().total_players;
    let should_add_player = {
        let game_rules = world.get_resource::<GameRules>().unwrap();
        let add_new_player = random::<bool>();
        add_new_player && total_players < game_rules.max_players
    };
    // Randomly add a new player
    if should_add_player {
        world.spawn().insert_bundle((
            Player {
                name: format!("Player {}", total_players),
            },
            Score { value: 0 },
        ));

        let mut game_state = world.get_resource_mut::<GameState>().unwrap();
        game_state.total_players += 1;
    }
}
```

but you'll notice that this code is dead. And when I attempt to do a similar form:

```
pub fn world_tick(world: &mut World) {
    // ...
}
```

... and add it to my App:

```
app.
            // ... snip
            .add_system(running_systems::world_tick)
            // ... snip
```

then I get a compile error that's pretty explicit but seems unfixable:

```
    Checking salamander v0.1.0 (/Users/richard.rast/rustdev/salamander)
error[E0599]: no method named `system` found for fn item `for<'r> fn(&'r mut bevy::prelude::World) {world_tick}` in the current scope
  --> src/main.rs:89:53
   |
89 |             .add_system(running_systems::world_tick.system())
   |                                                     ^^^^^^ method not found in `for<'r> fn(&'r mut bevy::prelude::World) {world_tick}`
   |
   = note: `running_systems::world_tick` is a function, perhaps you wish to call it
```

so maybe exclusive systems were broken by the system refactor with 0.6?

**Comment by rodya-mirov** [comment]
@alice-i-cecile Thanks! If I can get it working I can try and contribute a small example back to the repo. I think this is useful even without a "real, well-motivated" example, because for people like me, the [examples](https://github.com/bevyengine/bevy/tree/latest/examples) are the only useful documentation of most of bevy's actual API.

**Comment by alice-i-cecile** [comment]
That would be very much appreciated! I was thinking about this last night, and realized that this pattern may be useful for predictive AI. Perhaps a tic-tac-toe game example?

**Comment by mockersf** [comment]
if you don't want to go all the way to having your own schedule, you could also use run criteria, as they allow you to rerun systems inside the same frame. This is used for example in fixed time step to rerun systems in the same frame if the time step is smaller than the frame duration.

You can see it in action in the [`fixed_timestep.rs`](https://github.com/bevyengine/bevy/blob/main/examples/ecs/fixed_timestep.rs) example by setting it to a small time step (less than 0.016 on 60 fps)

**Comment by alice-i-cecile** [comment]
Yep, looping run criteria will also do the trick. They won't play nice with other run criteria (like states) though, and are likely to get cut as part of #2801 due to their sharp corners and excessive complexity :)

**Comment by rodya-mirov** [comment]
I think an exclusive system is probably what I want. I've already had to hack bevy more than I'd like to force all the systems to run in a **consistent** sequence (single threaded, apparently, is not enough) and manually clear events at the end of the tick so they're not processed twice. Clearly what I want is not what bevy is designed for but I think I've got the tools I need to make it work (still love the engine).

Which is to say, at this point I'd rather just assume direct control than try to solve my problems with configuration.
